### DISCUSSION TITLE
Roadmap comments from DAI perspective

### DISCUSSION THREAD

**Discussion by mmalohlava** [h2oai/datatable#2700] [General]
This is the list of issues annotated by the priority from DAI perspective:

## 🚨 Top Priority 
Goal: optimize computation speed, minimize data conversions

- [ ] Optimize MLI code to adopt datatable (DAI: <kbd>TOP</kbd>, time needed: ?)
- [x] Arrow support/conversion (TODO issue) (DAI: <kbd>TOP</kbd>, time needed: 2-3 weeks)
   - [x] Arrow -> datatable (#2711)
   - [x] datatable -> Arrow (#2717) (parallelized, there is still copy - see `to_arrow` [docs](https://datatable.readthedocs.io/en/latest/api/frame/to_arrow.html))
   - 📝 Note: the GPU version of XGB will benefit from it
   - ➡️ *Action*: @oleksiyskononenko @st-pasha needs to clarify usage (how to be memory efficient)
   - 📝  Note: Arrow -> datatable -> some limited ops (avoiding materialization) -> Arrow - will avoid copy
- [x] Date/time columns (#1646) (DAI: <kbd>TOP</kbd>, when available: End of February first phase - rough estimate)
  - [x] #2858: date32 column (date only);
  - [x] #2911: time64 column (timestamp).
- [x] Interpretable model (#2358) (DAI: <kbd>TOP</kbd>, when available: most of the features are already available [docs](https://datatable.readthedocs.io/en/latest/api/models/linear_model.html))
- [ ] Upgrade of datatable does not work with the latest DAI

## High Priority
- [ ] User-defined aggregator / map functions (#1960) (DAI: <kbd>HIGH</kbd>, time needed: ?)
- [ ] Tighter integration with LGBM (maybe the arrow support is the way): (DAI: <kbd>TOP</kbd>, time needed: ?) [clarify]
   - 📝 Note: LGBM can accept native datatable frame for computation (no conversion to pandas/numpy)
   - 📝 Note: This will need change in LGBM - since it needs numpy input (dups of memory, slow serial creation of data)
   - 📝 Note: It seems like LGBM does not support Arrow yet: https://github.com/microsoft/LightGBM/issues/2302
- [ ] Performance on multi-core systems/cache trashing (TODO issue) (DAI: <kbd>HIGH</kbd>, time needed: ?) [clarify]
- [ ] Fread improvements:
    - [ ] Implement streaming reading (#1958, ~#1843~, ~#1750~) (DAI: <kbd>HIGH</kbd>, time needed: ?)
      - @arnocandel we can pass total memory limitation to fread to enable chunk-based reading.

## Other
- [ ] Full support of datasets with >2B rows (#2336) (DAI: <kbd>MEDIUM</kbd>, time needed: ?) 
    - Note: ability to read and write is good enough as the first step
    - Note: efficient version needs lot of effort
- [ ] Inner/outer joins (#1080) (DAI: <kbd>LOW</kbd>)
- [ ] Ability to score dataframe using MOJO in efficient way (`f.score(mojo)`) (DAI: <kbd>MEDIUM</kbd>, time needed: ?) [clarify]
- [ ] Fread improvements:
    - [ ] Implement streaming reading (#1958, ~#1843~, ~#1750~) (DAI: <kbd>LOW</kbd>, time needed: ?)
    - [ ] Implement reading + filtering (#1950) (DAI: <kbd>LOW</kbd>, time needed: ?)
    - [ ] Better auto-detection of parse settings (~~#1343~~, #1045, #1036, #946, #934, #922, #518) (DAI: <kbd>LOW</kbd>, time needed: ?)

- [ ] Functions for string manipulations (DAI: <kbd>MEDIUM</kbd>, time needed: ?)
    - [x] Apply slice to a string (#1667)
    - [ ] Convert to upper/lower case
    - [ ] Repeat strings (`*`)
    - [ ] Strip whitespace / other characters
- [x] Functions for date/time manipulations (DAI: <kbd>LOW</kbd>, time needed: ?)
- [ ] Rolling windows support (#1500) (DAI: <kbd>LOW</kbd>, time needed: ?)

## New proposals
- [ ] Data compression (DAI: <kbd>MEDIUM</kbd>, time needed: ?)
- [ ] Data encryption in Jay file (DAI: <kbd>MEDIUM</kbd>, time needed: ?)
- [ ] Per column Jay file, support for efficient sporadic data updates (DAI: <kbd>LOW</kbd>, time needed: ?)
- [ ] Distributed mode of datatable (via Dask) (DAI: <kbd>LOW</kbd>, time needed: ?)

**Comment by mmalohlava** [comment]
Any news on time-estimates?

**Reply by oleksiyskononenko** [reply]
Yes, see some updates above. We also added a tag `[clarify]` for some items that are not fully clear from dt perspective at the moment.

**Reply by bramverbeke2** [reply]
Is there any idea when basic string manipulations would be possible (lower/upper)?

**Reply by oleksiyskononenko** [reply]
@bramverbeke2 This discussion was related to datatable improvements from the [DAI](https://www.h2o.ai/products/h2o-driverless-ai/) perspective. Therefore, the basic string manipulations were not a top priority. If you're interested in having the lower/upper case function, please open the corresponding issue.

**Comment by mmalohlava** [comment]
regarding:

> Tighter integration with LGBM (maybe the arrow support is the way):

My understanding: Right now, we load data in DTBL format, do munging, but before pushing into LGBM we need to convert data into format expected by LGBM (=copy of data). @arnocandel was proposing to integrate DTBL directly to LGBM to be able to accept data in DTBL format.

> Performance on multi-core systems/cache trashing 

@arnocandel ?

> Ability to score MOJO in efficient way

Able to use MOJO in _"map"_ call on data frame in a similar way how H2O-3/Spark are scoring with MOJO: split data to partitions and score partitions in parallel.

**Comment by mmalohlava** [comment]
**Actions**

- [ ] ➡️  @oleksiyskononenko @st-pasha needs to clarify usage (how to be memory efficient)
- [ ] ➡️  are there any PPC64le related issues (@surenH2oai asked)

**Comment by mmalohlava** [comment]
Current priority from H2O:
 - linear model - needs categorical support
    - action: needs meeting with @arnocandel @pseudotensor to figure out next steps
 - time column
 - flag for fread to disable datetime column
 - functionality of stream freading

From community perspective:
  - google summer of code - reshape/melt function
  - function apply
